### PR TITLE
[#6] [UI] As a user, I can see the Sign In page

### LINKED ISSUE #6: [UI] As a user, I can see the Sign In page
## Why
When the user opens the website, he should land on the Sign In page to sign in to use the application.

## Acceptance Criteria
- Create a Sign In page with the below components:
  - Email text field
  - Password text field
  - The Nimble logo
  - `Sign in to Nimble` text
  - Forgot? button
  - Submit button
  - The gradient background.
[Design](https://www.figma.com/file/i1AI0IzCWuJApX5CK0DmiW/Challenge---Web-App?type=design&node-id=6-168&t=9t3BfBxlZtL53kmn-0)
image

### PR DESCRIPTION
**Pull request by manh-t**
Close #6 

## What happened 👀

- Ignore `tailwind` in `stylelint`
- Remove sample code
- Create SignInScreen
- Write unittest.

## Insight 📝

I use `tailwind` to apply the styles to the components, and I create some reusable attributes to work with `tailwind`.

## Proof Of Work 📹

image

Check the Netlify link below.

### CODE REVIEW

**Comment by manh-t** [comment]
🚀 Deployed on https://649e50e8f120a02951095552--manh-react-survey.netlify.app

**Comment by manh-t** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Messages</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:book:</td>
      <td>Congrats, coverage is good</td>
    </tr>
  </tbody>
</table>

## Coverage in All Files
File | Line Coverage | Statement Coverage | Function Coverage | Branch Coverage
---- | ------------: | -----------------: | ----------------: | --------------:
[src/App.tsx](../blob/e277b2e/src/App.tsx) | (3/3) 100% | (3/3) 100% | (1/1) 100% | (0/0) NaN%
[src/index.tsx](../blob/e277b2e/src/index.tsx) | (3/3) 100% | (3/3) 100% | (0/0) NaN% | (0/0) NaN%
[src/../index.tsx](../blob/e277b2e/src/components/ElevatedButton/index.tsx) | (3/3) 100% | (3/3) 100% | (1/1) 100% | (0/0) NaN%
[src/../TextInput/index.tsx](../blob/e277b2e/src/components/TextInput/index.tsx) | (3/3) 100% | (3/3) 100% | (1/1) 100% | (2/2) 100%
[src/lib/requestManager.ts](../blob/e277b2e/src/lib/requestManager.ts) | (5/5) 100% | (5/5) 100% | (2/2) 100% | (1/1) 100%
[src/routes/index.tsx](../blob/e277b2e/src/routes/index.tsx) | (1/1) 100% | (1/1) 100% | (0/0) NaN% | (0/0) NaN%
[src/screens/SignIn/index.tsx](../blob/e277b2e/src/screens/SignIn/index.tsx) | (3/3) 100% | (3/3) 100% | (1/1) 100% | (0/0) NaN%
Total | (21/21) 100% | (21/21) 100% | (6/6) 100% | (3/3) 100%

<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against c5133e8423bb3bf6010e64ea3847e2b36a1878f7
</p>

File: .eslintrc.js [hunk 1]
```diff
@@ -1,5 +1,10 @@
 module.exports = {
-  extends: ['@nimblehq/eslint-config-nimble-react', '@nimblehq/eslint-config-nimble-testing'],
+  extends: [
+    '@nimblehq/eslint-config-nimble-react',
+    '@nimblehq/eslint-config-nimble-testing',
+    'plugin:cypress/recommended',
+    'prettier',
```
Comment (nvminhtue) [inline .eslintrc.js]: we can remove it as already added in `@nimblehq/eslint-config-nimble-react`, we don't need to override

File: src/components/ElevatedButton/index.tsx [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+  isFullSize: boolean;
+}
+
+const ElevatedButton = ({ children, isFullSize, ...attributes }: ElevatedButtonProps): JSX.Element => {
+  return (
+    <button
+      type="button"
+      className={
+        'bg-white text-black-chinese font-bold text-regular rounded-[10px] focus:outline-none focus:shadow-outline h-14' +
+        (isFullSize ? ' w-full' : '')
```
Comment (nvminhtue) [inline src/components/ElevatedButton/index.tsx]: let's try to concat by using `` string with dynamic variable

File: src/screens/SignIn/index.tsx [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+  passwordField: 'sign-in-form__input-password',
+  forgotButton: 'sign-in-form__forgot-button',
+  signInButton: 'sign-in-form__button',
+};
+
+const SignInScreen = (): JSX.Element => {
+  return (
+    <div
+      className="sign-in bg-cover min-h-screen flex flex-col justify-center items-center"
+      style={{ backgroundImage: 'url(' + background + ')' }}
```
Comment (nvminhtue) [inline src/screens/SignIn/index.tsx]:
```suggestion
      style={{ backgroundImage: `url '${background}'` }}
```

Comment (nvminhtue) [reply src/screens/SignIn/index.tsx]: But can we try [this](https://tailwindcss.com/docs/background-image) out?

File: src/screens/SignIn/index.tsx [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+      <form className="w-80" data-test-id={signInScreenTestIds.signInForm}>
+        <div className="mb-6">
+          <TextInput
+            label="Email"
+            labelDataTestId={signInScreenTestIds.emailLabel}
+            inputAttributes={{
+              id: 'form-sign-in-email',
+              required: true,
+              type: 'email',
+              'data-test-id': signInScreenTestIds.emailField,
```
Comment (nvminhtue) [inline src/screens/SignIn/index.tsx:32]:
```suggestion
              data-test-id: signInScreenTestIds.emailField,
```

**Comment by nvminhtue** [comment]
image
Could you help me to clarify it?

File: src/screens/SignIn/index.tsx (hunk 3 above)
Comment (manh-t) [reply src/screens/SignIn/index.tsx]: @nvminhtue ah thanks, I will update it 👏

File: src/screens/SignIn/index.tsx (hunk 4 above)
Comment (manh-t) [reply src/screens/SignIn/index.tsx:32]: @nvminhtue unfortunately, I cannot remove the single quotes due to the error 🙏 
image

**Comment by manh-t** [comment]
> image Could you help me to clarify it?

Thanks for catching the issue. This is the issue related to autofill. Fixed in [5c478f2](https://github.com/manh-t/react-survey/pull/42/commits/5c478f27a1d3505c73d5809f447bc9c1ba77a217) 🙏 
image

**Comment by nvminhtue** [comment]
The spacing seems closer compared to the design. Please take a look at that and appropriately adjust.
The font also different, when the one in the design is (`Neuzeit S LT Std`)
image

**Comment by manh-t** [comment]
> The spacing seems closer compared to the design. Please take a look at that and appropriately adjust. The font also different, when the one in the design is (`Neuzeit S LT Std`) image

I adjusted the spacing ✅ 

Regarding the font, the default font is currently `Neuzeit S LT Std`
image

But the letter spacing is different, so I added the letter spacing in 6be17d4
Left (design) vs Right (actual)
image

**Review by nvminhtue — APPROVED** [review]
👏

File: src/screens/SignIn/index.tsx [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+            />
+          </span>
+          <button
+            className="text-white text-opacity-50 text-small tracking-survey-normal absolute right-3 bottom-4"
+            data-test-id={signInScreenTestIds.forgotButton}
+          >
+            Forgot?
+          </button>
+        </div>
+        <ElevatedButton isFullSize={true} type="submit" data-test-id={signInScreenTestIds.signInButton}>
```
Comment (tyrro) [inline src/screens/SignIn/index.tsx]: attributes without values are interpreted as a boolean true, so we can write:
 
```suggestion
        <ElevatedButton isFullSize type="submit" data-test-id={signInScreenTestIds.signInButton}>
```

File: src/assets/stylesheets/application.scss [hunk 6]
```diff
@@ -1,3 +1,29 @@
-@import 'tailwindcss/base';
-@import 'tailwindcss/components';
-@import 'tailwindcss/utilities';
+@tailwind base;
+@tailwind components;
+@tailwind utilities;
```
Comment (tyrro) [inline src/assets/stylesheets/application.scss]: could we move the tailwind-specific lines to a separate file `base/tailwind.scss` and import it in `base`? The idea is to keep `application.scss` clear and concise.

File: src/components/ElevatedButton/index.tsx
Comment (tyrro) [inline src/components/ElevatedButton/index.tsx:1]: could we have unit tests for the components? Or do you plan to do it separately?

**Review by tyrro — CHANGES_REQUESTED** [review]

File: src/components/ElevatedButton/index.tsx
Comment (manh-t) [reply src/components/ElevatedButton/index.tsx:1]: @tyrro ah let me add the unittest for components 🙏

File: src/screens/SignIn/index.tsx (hunk 5 above)
Comment (manh-t) [reply src/screens/SignIn/index.tsx]: @tyrro resolved in [1aa1c5e](https://github.com/manh-t/react-survey/pull/42/commits/1aa1c5e5c8d8d2c9cda3ac696056fd9e24c867d1)

File: src/assets/stylesheets/application.scss (hunk 6 above)
Comment (manh-t) [reply src/assets/stylesheets/application.scss]: @tyrro resolved in [1aa1c5e](https://github.com/manh-t/react-survey/pull/42/commits/1aa1c5e5c8d8d2c9cda3ac696056fd9e24c867d1)

File: src/components/ElevatedButton/index.tsx
Comment (manh-t) [reply src/components/ElevatedButton/index.tsx:1]: @tyrro added in [1aa1c5e](https://github.com/manh-t/react-survey/pull/42/commits/1aa1c5e5c8d8d2c9cda3ac696056fd9e24c867d1), thanks a lot 😄

File: src/assets/stylesheets/application.scss [hunk 7]
```diff
@@ -1,3 +1,25 @@
...
+textarea:-webkit-autofill:focus,
+select:-webkit-autofill,
+select:-webkit-autofill:hover,
+select:-webkit-autofill:focus {
+  -webkit-text-fill-color: #fff;
+  /* -webkit-box-shadow: 0 0 0px 1000px #1b1b6c inset; */
+  transition: background-color 5000s ease-in-out 0s;
+
+  background-color: #fff;
+}
```
Comment (tyrro) [inline src/assets/stylesheets/application.scss]: could we move these to `tailwind.scss` as these are specific to tailwind?

File: src/assets/stylesheets/base/_tailwindcss.scss
Comment (tyrro) [inline src/assets/stylesheets/base/_tailwindcss.scss:1]: maybe we could just name this file to simply `_tailwind.scss`?

File: src/components/ElevatedButton/index.test.tsx [hunk 8]
```diff
@@ -0,0 +1,39 @@
+import React from 'react';
+
+import { render, screen } from '@testing-library/react';
+
+import ElevatedButton from '.';
+
+describe('ElevatedButton', () => {
+  const label = 'Sign in';
+  const dataTestId = 'sign-in_button';
```
Comment (tyrro) [inline src/components/ElevatedButton/index.test.tsx:9]: let's have an empty line before the start of  the `it`  block!

File: src/components/ElevatedButton/index.tsx [hunk 9]
```diff
@@ -0,0 +1,21 @@
...
+  children: React.ReactNode;
+  isFullSize: boolean;
+}
+
+const ElevatedButton = ({ children, isFullSize, ...attributes }: ElevatedButtonProps): JSX.Element => {
+  const fullWidth = isFullSize ? 'w-full' : '';
+  return (
+    <button
+      type="button"
+      className={`bg-white text-black-chinese font-bold text-regular tracking-survey-tight rounded-[10px] focus:outline-none focus:shadow-outline h-14 ${fullWidth}`}
```
Comment (tyrro) [inline src/components/ElevatedButton/index.tsx]: let's use [classnames](https://github.com/JedWatson/classnames) library here, it'll simplify using classnames a lot 😉 
For example, this code can be re-written as:
```suggestion
            className={classNames(
              'bg-white text-black-chinese font-bold text-regular tracking-survey-tight rounded-[10px] focus:outline-none focus:shadow-outline h-14',
              { 'w-full': isFullSize },
            )}
```

File: src/components/TextInput/index.test.tsx [hunk 10]
```diff
@@ -0,0 +1,52 @@
...
+
+import TextInput from '.';
+
+describe('TextInput', () => {
+  const label = 'Email';
+  const labelDataTestId = 'email-label';
+  const inputDataTestId = 'text-input';
+  const inputType = 'email';
+  const inputPlaceholder = 'input placeholder';
+  it('renders a text input with a label', () => {
```
Comment (tyrro) [inline src/components/TextInput/index.test.tsx:14]:
```suggestion
 
   it('renders a text input with a label', () => {
```

File: src/components/ElevatedButton/index.test.tsx [hunk 11]
```diff
@@ -0,0 +1,39 @@
...
+
+    const button = screen.getByTestId(dataTestId);
+
+    expect(button).toBeVisible();
+    expect(button).toHaveTextContent(label);
+    expect(button).toHaveAttribute('type', 'button');
+    expect(button).toHaveClass('w-full');
+  });
+
+  it('renders a wrap-content button', () => {
```
Comment (tyrro) [inline src/components/ElevatedButton/index.test.tsx]: we could simply say that

```suggestion
  it('does NOT render a full-width button', () => {
```

It makes it easier to read, what do you think?

File: src/components/ElevatedButton/index.tsx [hunk 12]
```diff
@@ -0,0 +1,21 @@
+import React from 'react';
+
+interface ElevatedButtonProps extends React.ButtonHTMLAttributes<HTMLButtonElement> {
+  children: React.ReactNode;
+  isFullSize: boolean;
```
Comment (tyrro) [inline src/components/ElevatedButton/index.tsx]: I am whether it'll be more appropriate if we name this prop `isFullWidth`?

File: src/components/TextInput/index.test.tsx [hunk 13]
```diff
@@ -0,0 +1,52 @@
...
+        id: 'sign-in-email',
+        required: true,
+        type: inputType,
+        placeholder: inputPlaceholder,
+        'data-test-id': inputDataTestId,
+      },
+    };
+    render(<TextInput {...inputProps} />);
+
+    expect(screen.queryByRole('label')).not.toBeInTheDocument();
```
Comment (tyrro) [inline src/components/TextInput/index.test.tsx:51]: [toBeInTheDocument](https://github.com/testing-library/jest-dom#tobeinthedocument) passes if the element is `display: none` or `visibility: hidden`, which we don't want in this case, so could we use `toBeVisible`?

```suggestion
    expect(screen.queryByRole('label')).not.toBeVisible();
```

**Review by tyrro — CHANGES_REQUESTED** [review]

File: src/screens/SignIn/index.tsx [hunk 14]
```diff
@@ -0,0 +1,65 @@
...
+            <TextInput
+              label="Password"
+              labelDataTestId={signInScreenTestIds.passwordLabel}
+              inputAttributes={{
+                id: 'form-sign-in-password',
+                required: true,
+                type: 'password',
+                'data-test-id': signInScreenTestIds.passwordField,
+              }}
+              extraClassName="pr-16"
```
Comment (tyrro) [inline src/screens/SignIn/index.tsx]: we could just name it as `className`,  it's a common naming convention, for example in React codebase:

https://github.com/facebook/react/blob/a1c62b8a7635c0bc51e477ba5437df9be5a9e64f/packages/react-devtools-shared/src/devtools/views/Icon.js#L32

File: tailwind.config.js [hunk 15]
```diff
@@ -2,7 +2,29 @@
...
+      },
+      fontSize: {
+        small: ['15px', '20px'],
+        regular: ['17px', '22px'],
+      },
+      backgroundImage: {
+        'sign-in': "url('assets/images/illustrations/background-sign-in.png')",
+      },
+      letterSpacing: {
+        'survey-tighest': '-1px',
```
Comment (tyrro) [inline tailwind.config.js:20]: hopefully, we don't have to invent any more `tight`-adjectives lol 😆 
btw  the naming [goes with the tailwind world](https://tailwindcss.com/docs/letter-spacing) as well 👍🏿

**Review by tyrro — CHANGES_REQUESTED** [review]

File: src/components/ElevatedButton/index.tsx (hunk 9 above)
Comment (tyrro) [reply src/components/ElevatedButton/index.tsx]: Also, we could extract the class names to a variable for better readability:

```js
const DEFAULT_CLASS_NAMES = 'bg-white text-black-chinese font-bold text-regular tracking-survey-tight rounded-[10px] focus:outline-none focus:shadow-outline h-14'
```

**Review by tyrro — CHANGES_REQUESTED** [review]

File: src/components/ElevatedButton/index.tsx (hunk 9 above)
Comment (manh-t) [reply src/components/ElevatedButton/index.tsx]: TIL, thanks, @tyrro

File: src/components/ElevatedButton/index.tsx (hunk 12 above)
Comment (manh-t) [reply src/components/ElevatedButton/index.tsx]: @tyrro make sense, let me update it 🙏

File: src/screens/SignIn/index.tsx (hunk 14 above)
Comment (manh-t) [reply src/screens/SignIn/index.tsx]: got it @tyrro 🚀

File: tailwind.config.js (hunk 15 above)
Comment (manh-t) [reply tailwind.config.js:20]: @tyrro I checked the Figma and added all customized letter spacings here, so that's all 😄

File: src/components/TextInput/index.test.tsx (hunk 13 above)
Comment (manh-t) [reply src/components/TextInput/index.test.tsx:51]: @tyrro actually when the label is empty, the `<label>` element will not be generated so `toBeInTheDocument()` is my expected result in this case 😄 

I also try to use `toBeVisible()` but it doesn't work with the does-not-exist element 🙏 
image

File: src/assets/stylesheets/application.scss (hunk 7 above)
Comment (manh-t) [reply src/assets/stylesheets/application.scss]: @tyrro resolved in [a31c486](https://github.com/manh-t/react-survey/pull/42/commits/a31c486e6aabdb861be574b7d76fd7ec5686a78d)

File: src/components/ElevatedButton/index.test.tsx (hunk 8 above)
Comment (manh-t) [reply src/components/ElevatedButton/index.test.tsx:9]: @tyrro resolved in [a31c486](https://github.com/manh-t/react-survey/pull/42/commits/a31c486e6aabdb861be574b7d76fd7ec5686a78d)

File: src/components/ElevatedButton/index.tsx (hunk 12 above)
Comment (manh-t) [reply src/components/ElevatedButton/index.tsx]: @tyrro resolved in [a31c486](https://github.com/manh-t/react-survey/pull/42/commits/a31c486e6aabdb861be574b7d76fd7ec5686a78d)

File: src/screens/SignIn/index.tsx (hunk 14 above)
Comment (manh-t) [reply src/screens/SignIn/index.tsx]: @tyrro resolved in [a31c486](https://github.com/manh-t/react-survey/pull/42/commits/a31c486e6aabdb861be574b7d76fd7ec5686a78d)

File: src/components/ElevatedButton/index.tsx [hunk 16]
```diff
@@ -0,0 +1,20 @@
...
+
+interface ElevatedButtonProps extends React.ButtonHTMLAttributes<HTMLButtonElement> {
+  children: React.ReactNode;
+  isFullWidth: boolean;
+}
+
+const ElevatedButton = ({ children, isFullWidth, ...attributes }: ElevatedButtonProps): JSX.Element => {
+  const DEFAULT_CLASS_NAMES =
+    'bg-white text-black-chinese font-bold text-regular tracking-survey-tight rounded-[10px] focus:outline-none focus:shadow-outline h-14';
+  return (
```
Comment (tyrro) [inline src/components/ElevatedButton/index.tsx:14]: let's have an empty line before the `return` statement 🙌🏿

File: src/components/TextInput/index.test.tsx (hunk 13 above)
Comment (tyrro) [reply src/components/TextInput/index.test.tsx:51]: ah I see, we are testing a non-existing element!

File: src/components/TextInput/index.tsx [hunk 17]
```diff
@@ -0,0 +1,41 @@
...
+    placeholder?: string;
+    'data-test-id'?: string;
+  };
+  classNames?: string;
+};
+
+const TextInput = ({ label, labelDataTestId, inputAttributes, classNames }: TextInputProps): JSX.Element => {
+  const DEFAULT_CLASS_NAMES =
+    'block appearance-none bg-white bg-opacity-[.18] rounded-[12px] w-full h-14 focus:outline-none focus:ring-transparent px-3 mt-2 text-white text-regular tracking-survey-tight focus:bg-opacity-30';
+  return (
```
Comment (tyrro) [inline src/components/TextInput/index.tsx:22]: ditto about the empty line!

File: src/components/TextInput/index.tsx [hunk 18]
```diff
@@ -0,0 +1,41 @@
+import React from 'react';
+
+import classnames from 'classnames';
```
Comment (tyrro) [inline src/components/TextInput/index.tsx]: let's import `classNames`, it's similar to `classname` but goes with the camel case naming of JS! 😄 

```suggestion
import classNames from 'classnames';
```

File: src/screens/SignIn/index.tsx [hunk 19]
```diff
@@ -0,0 +1,65 @@
+import React from 'react';
+
+import nimbleLogoWhite from 'assets/images/icons/nimble-logo-white.svg';
```
Comment (tyrro) [inline src/screens/SignIn/index.tsx:3]: I think we are going to import a lot of assets in the coming days:
what about we set the path in `tsconfig.json` so that we can import it like this:

`import nimbleLogoWhite from 'icons/nimble-logo-white.svg';`

```json
"paths": {
        "icons/*": ["assets/images/icons*"]
 }
 ```
 
 You can do it in a separate PR if you want as you probably need to update `jest.config.js` as well!

**Review by tyrro — APPROVED** [review]
I've some last comments but the changes now LGTM! Well done!

File: src/screens/SignIn/index.tsx (hunk 19 above)
Comment (manh-t) [reply src/screens/SignIn/index.tsx:3]: I have tried this but getting error when running the application 😞 I will update it in [another ticket](https://github.com/manh-t/react-survey/issues/60) 🙏 
image
